### ISSUE TITLE
Edit ignores WFS layer's projection when it reads and writes features

### ISSUE THREAD

**Issue by jacobwod** [hajkmap/Hajk#927]
# Describe the bug
Since #884 the Edit plugin expects the WFS layers to be in the same projection as the View.

# To Reproduce
## Preparations - setup the map, WFS and WMS layers:

1. Setup a map in one projection (e.g. EPSG:3008)
![Skärmavbild 2021-11-23 kl  10 10 15](https://user-images.githubusercontent.com/110222/142997153-58de5369-935b-4830-87bd-f35364c80325.png)
1. Setup a Edit service (WFS) in another projection (e.g. EPSG:3006)
![Skärmavbild 2021-11-23 kl  10 11 10](https://user-images.githubusercontent.com/110222/142997299-4c4200e6-81f8-46ee-836e-e12e6bd93c90.png)
1. Setup _the same layer_ as a regular WMS layer (so we can view it via LayerSwitcher). Use the same projection as WFS (OpenLayes will reproject it to the View's projection)
![Skärmavbild 2021-11-23 kl  10 11 55](https://user-images.githubusercontent.com/110222/142997402-83a43566-bb7b-4823-a977-127ba07755fe.png)

## Preview the layer using WMS and PostGIS
Open the map in Client UI. Show the WMS layer. In my example, the layer has 4 geometries. They're in 3006 but get correctly placed (confirmed both in Hajk and pgAdmin):

![Skärmavbild 2021-11-23 kl  10 15 27](https://user-images.githubusercontent.com/110222/142997932-6181d6b0-9d3c-4dac-bf94-cc7348d83fc9.png)

![Skärmavbild 2021-11-23 kl  10 16 31](https://user-images.githubusercontent.com/110222/142998085-483ea602-88be-4c31-98c6-ce591146ed1d.png)

## Preview using WFS
Now try opening the WFS service in Edit plugin. The layer is fetched correctly, Hajks specifies that we want 3006 (that's the layer's projection):
```
GET https://myurl.com?service=WFS&version=1.1.0&request=GetFeature&typename=sitac_angrepp&srsname=EPSG:3006
```
And we do get back a valid GML response with geometries in 3006:

![bild](https://user-images.githubusercontent.com/110222/142998846-63e03cf4-7145-448c-bdcd-d606d6d04626.png)

**In spite of the correct response, the geometries are somehow lost (they're added to the View as if they were in the Views projection, 3008, and end up way outside their correct placement [which should be exactly as the WMS service returned]). This is the number 1 issue.**

## Add a new feature using WFS Edit
There is also a number 2 issue: as we've seen in issue number 1, the Edit plugin seems to ignore the WFS service's projection (3006) and just expects that the geometries are in View's projection (3008). So the four existing features won't show. This error works the other way around too: **when we add a geometry using Hajk, its coordinates will be read as if they were in the Views projection (3008), but saved in the layer's projection (3006), without any transformation occuring. Our new feature ends up in an unexpected place.**

Let's draw a feature. It will be placed just between 3 of the original features:
![Skärmavbild 2021-11-23 kl  10 26 57](https://user-images.githubusercontent.com/110222/142999572-a4755a37-f37e-4f5b-8e7b-cff577489f5d.png)

The POST looks fine, we want the geometry to be saved in 3006, which is the WFS service's projection and it's correct:
```xml
<Transaction
	xmlns="http://www.opengis.net/wfs" service="WFS" version="1.1.0" xsi:schemaLocation="http://www.opengis.net/wfs http://schemas.opengis.net/wfs/1.1.0/wfs.xsd"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<Insert>
		<sitac_angrepp
			xmlns="https://rtj.halmstad.se">
			<geometry>
				<LineString
					xmlns="http://www.opengis.net/gml" srsName="EPSG:3006">
					<posList srsDimension="2">123497.55967540591 6286955.857225036 119521.55967540592 6281523.857225036 126577.55967540591 6281243.857225036 123833.55967540591 6285611.857225036</posList>
				</LineString>
			</geometry>
			<kategori>Åtgärder</kategori>
			<typ>Angrepp</typ>
			<beskrivning>Genomfört direkt angrepp</beskrivning>
		</sitac_angrepp>
	</Insert>
</Transaction>
```

I can see the new geometry in PostGIS too:
![Skärmavbild 2021-11-23 kl  10 30 10](https://user-images.githubusercontent.com/110222/143000011-99b3e6c4-b110-4774-a7a7-de493fd38e81.png)

**But Hajk took the 3008 coordinates and just saved them into the 3006 layer, without transformation. This is the issue number 2.**

# Conclusion
I don't feel comfortable rolling out 3.8 with Edit being broken like this, what's your opinion @jesade-vbg @hitomin21?

**Comment by jacobwod** [comment]
# Tests
_Below is a comprehension of Edit functionality using our configuration between v3.7.0 and v3.8.0-rc.1._

## Hajk 3.7.0

### Scenario 1 - WFS from GeoServer, map and WFS layer in the same projection
In the video below I demonstrate how both add, update and delete work flawlessly.

https://user-images.githubusercontent.com/110222/143398461-c24abb51-b3e8-4283-8af7-b5af36fffa2d.mov

### Scenario 2 - WFS from QGIS Server, map and WFS layer in the same projections
In the video below data served from QGIS Server can be added and removed flawlessly and is instantaneously updated in the WMS service too. _The update however doesn't work, the geometry is lost._ 

https://user-images.githubusercontent.com/110222/143400069-c2e61981-a8e2-458e-b061-1c848bfc9195.mov

## Hajk 3.8.0-rc1+ (commit 0b72207a9304bd2806ba1b723363bd283d7bd503)

### Scenario 1 - WFS from GeoServer, map and WFS layer in the same projection
In the video below I demonstrate how both add, update and delete work flawlessly.

https://user-images.githubusercontent.com/110222/143402441-8a60c277-4471-429c-b02b-bc8bbfb982d8.mov

### Scenario 2 - WFS from QGIS Server, map and WFS layer in the same projections
Just like in 3.7, the update operation moves the geometry to an unknown location. Other than that it seems to work well, **as long as both the map and WFS layers are in the same SRS**.
https://user-images.githubusercontent.com/110222/143402158-4b5db922-48b3-435c-b2b4-b9898bd5c9f0.mov

# Conclusion

**Comment by hitomin21** [comment]
We are discussing a bit internally about this, we will get back to you @jacobwod. Thanks for a very detail description!

**Comment by hitomin21** [comment]
I think the simplest solution would be to use the view's projection in the request to the WFS and only use that in the edit tool.

At least Geoserver (and I assume QGIS Server as well) converts the response to the projection specified in the request so this would save us some trouble since we do not have to handle it in the code.

@jacobwod what do you think?

**Comment by jacobwod** [comment]
I think that we'll leave it _for now_. 

**This is still one of the major drawbacks in our current Edit plugin, however, and should be taken care of.**

**Comment by jacobwod** [comment]
@hitomin21 @jesade-vbg What happens when you try to edit/move an existing feature that has some `null` values in the database? (It's a completely valid use case.)

This is what I see:
![Skärmavbild 2021-11-30 kl  10 45 28](https://user-images.githubusercontent.com/110222/144023857-1c9b329b-9689-4d39-a668-a22c0347e55d.png)

**Comment by Hallbergs** [comment]
> @hitomin21 @jesade-vbg What happens when you try to edit/move an existing feature that has some `null` values in the database? (It's a completely valid use case.)
> 
> This is what I see: ![Skärmavbild 2021-11-30 kl 10 45 28](https://user-images.githubusercontent.com/110222/144023857-1c9b329b-9689-4d39-a668-a22c0347e55d.png)

The code below is probably the cause. `let value = field.defaultValue;` might be undefined? I might be wrong though.
```javascript
updateFeature() {
    const featureProps = this.props.model.editFeature.getProperties();
    Object.keys(this.state.formValues).forEach((key) => {
      let value = this.state.formValues[key];
      if (value === "") value = null;
      if (Array.isArray(value)) {
        value = value
          .filter((v) => v.checked)
          .map((v) => v.name)
          .join(";");
      }
      featureProps[key] = value;
    });

    this.props.editSource?.nonEditableFields?.forEach((field) => {
      let value = field.defaultValue;
      if (value === "") value = null;
      if (Array.isArray(value)) {
        value = value
          .filter((v) => v.checked)
          .map((v) => v.name)
          .join(";");
      }
      let geomName = this.props.model.editFeature.getGeometryName();
      if (!geomName) {
        geomName = "geom";
      }
      if (field.name !== geomName) {
        // should not overwrite the feature's geom
        featureProps[field.name] = value;
      }
    });
    this.props.model.editFeature.setProperties(featureProps);
  }
```

**Comment by jacobwod** [comment]
That's right, `featureProps[key] = value;` sets the value to `undefined` which is an Object, and upon writing as a String the default `toString()` method is invoked which leads to `"[Object object]"`.

![Skärmavbild 2021-11-30 kl  11 00 42](https://user-images.githubusercontent.com/110222/144026251-1adee867-d8bc-4768-8470-7b489b7cd78c.png)

Please fix this as soon as possible @hitomin21, @jesade-vbg so @Hallbergs and I can prepare a RC2.

**Comment by hitomin21** [comment]
I'm working on that fix now, thanks for the info.

**Comment by hitomin21** [comment]
let value = field.defaultValue

Would only be executed for noneditable fields so the bug must be in both noneditable and editable fields. I've added a check for if value is undefined. 

I also added so the default value (or empty string) will be added if we load a editable field with a null value (in initFormValues).

I'm still unable to reproduce the error with Geoserver so can't verify my fixes, sorry for making you re-test this over and over.

Pushed in branch https://github.com/hajkmap/Hajk/tree/fix/3.8-edittool
